### PR TITLE
update

### PR DESCRIPTION
**Pull request by Yolanda18-star**
@langyuemeng @YiniXu9506 @gengliqi PTAL

### CODE REVIEW

**Comment by langyuemeng** [comment]
LGTM

File: tikv-source-code-reading-21.md [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+
+我们希望 Merge 的设计能够满足以下几个需求：
+
+1. 不依靠 PD 在 Merge 期间不发起其他的调度满足正确性
+
+2. 不会因为网络隔离或者宕机引发正确性问题
+
+3. 只要参与 Merge 的 TiKV 中 Majority 存活且能互相通信，Merge 可以继续或者回滚，不会被阻塞住（跟 Raft 保证可用性的要求一致）
+
+4. 不对 Split/Conf Change 加额外条件限制（出于性能考虑）
```
Comment (gengliqi) [inline tikv-source-code-reading-21.md:27]:
```suggestion
4. 出于性能考虑，不对 Split/Conf Change 加额外条件限制
```

File: tikv-source-code-reading-21.md [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+2. 不会因为网络隔离或者宕机引发正确性问题
+
+    - 满足
+
+3. 只要参与 Merge 的 TiKV 中 Majority 存活且能互相通信，Merge 可以继续或者回滚，不会被阻塞住（跟 Raft 保证可用性的要求一致）
+
+    - 满足
+
+4. 不对 Split/Conf Change 加额外条件限制（出于性能考虑）
```
Comment (gengliqi) [inline tikv-source-code-reading-21.md:263]:
```suggestion
4. 出于性能考虑，不对 Split/Conf Change 加额外条件限制
```

File: tikv-source-code-reading-21.md [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+
+据此，可推出性质 2：Snapshot 不用考虑多次覆盖的 Merge 的情况，即 A merge to B，B merge to C（当然 A merge to B，C merge to B 的情况还是需要考虑的）。
+
+这一性质很好证明，如果某个隔离的 TiKV 上有 A B C 三个 Region 的 Peer，被隔离时，A merge to B 可以成功，但是 B merge to C 会被上述的检查条件 3 给挡住。
+
+有了该性质之后，Snapshot 只需要在 Apply 前删除那些要 Merge 给自己的 Source Peer 即可。但是删除这些 Source Peer 并不能随便删，必须要确保之后 Apply Snapshot 不会失败才能删。这里隐藏着一个违反性质 1 的陷阱，发送 Snapshot 并不意味着之后一定都会发 Snapshot，由于可能只有 Leader 把日志给 Compact 了，可以构造一种情况在 Leader 给某 Peer 发送 Snapshot 之后，之后的 Leader 仍然给它发送日志，如果此时 Source Peer 已经被删除，那就无法 Apply `CommitMerge` 了，违反性质 1。
+
+- 如何构造的问题就留给读者解答了。
+
+在 TiKV 的实现中是先检查该 Snapshot 是否在删除 Source Peer 之后范围不冲突（代码见 PeerFsmDelegate `check_snapshot`），然后在暂停 Source Peer 运行之后，对 Source Peer RegionLocalState 中的 PeerState 设置为 Tombstone，Target Peer RegionLocalState 中的 PeerState 设置为 Applying（代码见 PeerStorage `apply_snapshot`），并且使用 WriteBatch 原子的写入来解决该问题。
```
Comment (gengliqi) [inline tikv-source-code-reading-21.md:193]:
```suggestion
在 TiKV 的实现中是先检查该 Snapshot 是否在删除 Source Peer 之后范围不冲突（代码见 PeerFsmDelegate `check_snapshot`），然后在暂停 Source Peer 运行之后，对 Source Peer RegionLocalState 中的 PeerState 设置为 Tombstone，Target Peer RegionLocalState 中的 PeerState 设置为 Applying（代码见 PeerStorage `apply_snapshot`），并且使用 Rocksdb WriteBatch 原子的写入来解决该问题。
```

File: tikv-source-code-reading-21.md [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+Region C 进行了 2 次 Merge 以及 Split 之后，再次进行 Conf Change，移除了 TiKV 4 上的 Peer，在 TiKV 1 上增加 Peer。
+
+|  TiKV  |  conf change  |
+|  ----  | ----  |
+|  1  | A A C  |
+|  2  | D E C  |
+|  3  | D E C  |
+|  4  | D E *  |
+
+此时已经打破前提了，在 TiKV 1 上，Region C 已经在 Region A Apply Split 之前出现了。
```
Comment (gengliqi) [inline tikv-source-code-reading-21.md:241]:
```suggestion
此时已经打破前提了，在 TiKV 1 上，Region C 已经在 Region A Apply `BatchSplit` 之前出现了。
```

File: tikv-source-code-reading-21.md [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+
+6. 尽量减少 Merge 期间服务不可用的时间
+
+    - Target Region 无服务不可用时间
+
+    - Source Region 服务不可用时间如下
+
+      ![1-unavailabletime](media/tikv-source-code-reading-21/1-unavailabletime.png)
+
+    - PD 会选择较冷的 Region 作为 Source Region，所以一般来说这个代价是可忍受的
```
Comment (gengliqi) [inline tikv-source-code-reading-21.md:279]:
```suggestion
    - PD 会选择较冷的 Region 作为 Source Region，所以一般来说这个代价是可承受的
```

File: tikv-source-code-reading-21.md (hunk 3 above)
Comment (gengliqi) [reply tikv-source-code-reading-21.md:193]: sorry, should be `RocksDB`

**Review by gengliqi — APPROVED** [review]
LGTM

**Review by YiniXu9506 — APPROVED** [review]
LGTM

**Pull request merged**